### PR TITLE
Timesheets general style improvals - Timesheet autosave functionality fix - Timesheet autodelete functionality implemented

### PR DESCRIPTION
**Pull request by FernandoAurelius**
### **Pull Request Description**

This PR introduces the following main changes:

#### **1. Navigation Between Employees of the Day**
- **Added buttons** to make navigation between employees of the day easier.
- Changes made in:
  - `time_entry_view.py`.
  - Related templates.
- **Functionality:** Load more information between URLs and contexts.

#### **2. Bug Fixes in AG Grid**
- Resolved an issue where **entries were not being automatically saved** when created using AG Grid shortcuts.

#### **3. Auto-Deletion of Entries**
- Implemented an **auto-delete function** for entries removed via the trash can button in AG Grid.

#### **4. Efficient Notification System**
- Created a notification system with JavaScript to:
  - Send and receive AJAX requests with Django messages from views, and load the messages on the screen.
  - Changed `time_entry_view.py`, `timesheet_entry.js`, `timesheet_autosave.js` and created `messages.js`.

---

### **Additional Improvements**

- **Entry Creation Form:**
  - Created to better understand how the timesheet entry model works.
- **Styling Updates on the Employee Listing Page:**
  - Updated to use a Bootstrap-styled list.

### CODE REVIEW

**Review by corrin — COMMENTED** [review]
Hey, looks like this should be split over two PRs, there's a bunch of docs/general stuff, and then changes to time entry.

Lets' keep it at one for now but in the future, try and keep one ticket per PR

File: workflow/forms.py [hunk 1]
```diff
@@ -62,27 +62,55 @@ class Meta:
         exclude = ["job_pricing"]
 
 
-class TimeEntryForm(forms.ModelForm):
+class TimeEntryForm(forms.ModelForm):                                                                      # Form for creating and editing TimeEntry instances
```
Comment (corrin) [inline workflow/forms.py:65]: I don't see hte point of this comment.

A class called TimeEntryForm is obviously for time entries.

Let's keep comments to WHY things are like they are, not rewriting the code in english.

File: workflow/forms.py [hunk 2]
```diff
@@ -62,27 +62,55 @@ class Meta:
...
     class Meta:
         model = TimeEntry
-        widgets = {
-            "date": forms.DateInput(attrs={"type": "date"}),  # Keeps the date picker
+
```
Comment (corrin) [inline workflow/forms.py:68]: Again, this isn't adding anything.

File: workflow/forms.py [hunk 3]
```diff
@@ -62,27 +62,55 @@ class Meta:
...
+            "wage_rate",
+            "charge_out_rate",
+        )
 
-        # Drop-down for staff
-        self.fields["staff"].queryset = Staff.objects.all()
+    def __init__(self, *args, staff_member=None, **kwargs):                                                # Initialize the form with optional staff_member paramete
+        super().__init__(*args, **kwargs)
+        if staff_member:
+            # Handle case when staff_member is provided
```
Comment (corrin) [inline workflow/forms.py:89]: Again.

Anyone reading the code can read code.  DOn't rewrite it in English.

Explain your thought process.  So in this case .... why have you decided to handle having a staff member vs not differently?  If that's worth commenting on then comment on that.

File: workflow/forms.py [hunk 4]
```diff
@@ -62,27 +62,55 @@ class Meta:
...
+            self.fields["staff"].initial = staff_member                                                    # Pre-select the staff field with provided staff member
```
Comment (corrin) [inline workflow/forms.py:90]: I'm repeating myself.  This is why it's good to catch up more frequently - it cuts down on making the same comment repeatedly :)

File: workflow/forms.py [hunk 5]
```diff
@@ -62,27 +62,55 @@ class Meta:
...
+            self.fields["staff"].widget.attrs['disabled'] = True                                           # Disable staff field to prevent changes
+            self.fields["staff"].required = False                                                          # Make staff field not required as it's set automatically
+
+            # Pre-populate rate fields
+            self.fields["wage_rate"].initial = staff_member.wage_rate                                      # Set initial wage rate from staff member
+            # self.fields["charge_out_rate"].initial = staff_member.charge_out_rate                        # Set initial charge-out rate from staff member - currently not working because TimeEntry model has not charge_out_rate property
```
Comment (corrin) [inline workflow/forms.py:96]: This comment ... REALLY isn't helpful

File: workflow/forms.py [hunk 6]
```diff
@@ -62,27 +62,55 @@ class Meta:
...
+        else:
+            # Handle case when no staff_member is provided
+            self.fields["staff"].queryset = Staff.objects.all()                                            # Load all staff members for selection
+
+            # Handle rate population for existing instances
+            if "instance" in kwargs and kwargs["instance"] and kwargs["instance"].staff:                   # Check if editing existing TimeEntry
+                staff = kwargs["instance"].staff
+                self.fields["wage_rate"].initial = staff.wage_rate                                         # Set wage rate from existing entry's staff
+                '''try:                                                                                    # Safely handle charge_out_rate for existing instances
```
Comment (corrin) [inline workflow/forms.py:105]: But staff don't have a charge out rate?

File: workflow/models/time_entry.py [hunk 7]
```diff
@@ -69,8 +69,8 @@ class Meta:
         ordering = ["created_at"]  # Default ordering by creation time
 
     def save(self, *args, **kwargs):
-        # Ensure minutes and hours stay consistent
-        if self.items is not None and self.minutes_per_item is not None:
+        # Only calculate hours from items/minutes if hours isn't explicitly set
```
Comment (corrin) [inline workflow/models/time_entry.py:72]: Why?  

Again.  Do not rewrite code in comments

File: workflow/static/js/timesheet_autosave.js [hunk 8]
```diff
@@ -1,4 +1,13 @@
-// Debounce function to reduce frequent server calls
+let deletedEntries = [];
+
+function markEntryAsDeleted(entryId) {
+    if (entryId) {
```
Comment (corrin) [inline workflow/static/js/timesheet_autosave.js:4]: Ok.

I thought we were simply getting the data from the form and essentially replacing the data in the data model from the form?

File: workflow/templates/base.html [hunk 9]
```diff
@@ -7,6 +7,7 @@
     <meta name="viewport" content="width=device-width, initial-scale=1.0">
     <title>Jobs Manager</title>
     <link rel="icon" href="{% static 'favicon.ico' %}" type="image/x-icon">
+    <script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
```
Comment (corrin) [inline workflow/templates/base.html:10]: No, need to follow the pattern

File: workflow/templates/time_entries/timesheet_form.html [hunk 10]
```diff
@@ -0,0 +1,19 @@
+{% load crispy_forms_tags %}
+
```
Comment (corrin) [inline workflow/templates/time_entries/timesheet_form.html:2]: What's this file for?

File: workflow/views/time_overview_view.py [hunk 11]
```diff
@@ -32,6 +32,8 @@ def get(self, request, start_date=None, *args, **kwargs):
         staff_data = []
         all_staff = Staff.objects.all()
         for staff_member in all_staff:
+            if staff_member.is_staff is True:
```
Comment (corrin) [inline workflow/views/time_overview_view.py:35]: This is a perfect example of where comments are important.

Something like "App users are not workers in the factory and so should not be having timesheet entries"

**Review by corrin — CHANGES_REQUESTED** [review]

**Comment by FernandoAurelius** [comment]
In some files, I tried to be very explicit about what the code does rather than just explaining its intent, with the main goal of getting more familiar with the pre-existing codebase, but I will make the requested corrections by adopting a context-oriented documentation approach.

**Comment by FernandoAurelius** [comment]
> Hey, looks like this should be split over two PRs, there's a bunch of docs/general stuff, and then changes to time entry.
> 
> Lets' keep it at one for now but in the future, try and keep one ticket per PR

I think I merged docs into this branch after I opened the PR to docs, so the branch itself shows all the differences including the docs that I had previously pulled, but I only changed the timesheet related code itself in my work on the branch.

File: workflow/models/time_entry.py (hunk 7 above)
Comment (FernandoAurelius) [reply workflow/models/time_entry.py:72]: I updated all the comments you mentioned to improve code readability while also explaining my intention in the sections where I thought it made sense to do this.

Comment (FernandoAurelius) [reply workflow/models/time_entry.py]: Also, in the main changes I made and in the new functions / functions I changed a lot I decided to add more detailed documentation (especially in the JS functions), but still explaining the thought process / business logic / intention of the code.

**Comment by FernandoAurelius** [comment]
All requested changes made.

**Review by corrin — APPROVED** [review]

**Pull request merged**